### PR TITLE
Fix fuzz_level related thing, separate on_replace/on_remove from Scheduler & various fixes

### PR DESCRIPTION
**Pull request by tokatoka**
- Rename fuzz_level to scheduled_count
- I forgot to increment fuzz_level (scheduled_count) before, so fix this
- Add set_current_scheduled method. scheduler->next() calls this. it sets corpus.current() and update scheduled_count.
- Separate on_remove/on_replace from Scheduler. AFL never expects any corpus entry. so not every scheduler can have this.
- Remove HasFuzzedCorpusId.
- Remove fuzzed member from Testcase. fuzz_level > 0 is the equivalent thing

### CODE REVIEW

File: libafl/src/fuzzer/mod.rs [hunk 1]
```diff
@@ -387,7 +384,6 @@ where
 
                 // Add the input to the main corpus
                 let mut testcase = Testcase::with_executions(input.clone(), *state.executions());
-                testcase.set_parent_id_optional(state.fuzzed_corpus_id());
```
Comment (domenukk) [inline libafl/src/fuzzer/mod.rs:390]: Set the .current ID here

Comment (tokatoka) [reply libafl/src/fuzzer/mod.rs:390]: i'll do it on_add

File: libafl/src/schedulers/mod.rs [hunk 2]
```diff
@@ -104,16 +114,38 @@ impl<S> Scheduler for RandScheduler<S>
...
     S: HasCorpus + HasRand,
 {
+    fn on_add(&mut self, state: &mut Self::State, idx: CorpusId) -> Result<(), Error> {
+        // Set parent id
+        let current_idx = *state.corpus().current();
+        state
+            .corpus()
+            .get(idx)?
+            .borrow_mut()
+            .set_parent_id_optional(current_idx);
```
Comment (tokatoka) [inline libafl/src/schedulers/mod.rs:124]: @domenukk 
setting parent id here.

**Comment by andreafioraldi** [comment]
This field is also doing the same thing as fuzz_level > 0 https://github.com/AFLplusplus/LibAFL/blob/main/libafl/src/schedulers/ecofuzz.rs#L50

File: libafl/src/schedulers/mod.rs (hunk 2 above)
Comment (domenukk) [reply libafl/src/schedulers/mod.rs:124]: Is this even called for Objectives?

File: libafl/src/fuzzer/mod.rs (hunk 1 above)
Comment (domenukk) [reply libafl/src/fuzzer/mod.rs:390]: Don't think that's correct, see my comment below (probably doesn't work for Objectives as they are a different corpus(?)

Comment (domenukk) [reply libafl/src/fuzzer/mod.rs:390]: @tokatoka ?

Comment (tokatoka) [reply libafl/src/fuzzer/mod.rs:390]: yeah was busy last two days.
now I put this back

File: libafl/src/schedulers/mod.rs (hunk 2 above)
Comment (tokatoka) [reply libafl/src/schedulers/mod.rs:124]: No. i put set_parent_id_optional back for objectives

**Comment by tokatoka** [comment]
@andreafioraldi 
I changed on_evaluation to on_execution and made it to get called right after execution.

Else we can't compute the hash of the raw map when hitcount observer is used

**Comment by andreafioraldi** [comment]
Isn't it the opposite? on_evaluation ensure that observer.post_exec was called, on_execution no

**Comment by tokatoka** [comment]
yeah but we want to compute the hash before it is called.

Here in AFL++.
hash is computed here on line 481
https://github.com/AFLplusplus/AFLplusplus/blob/stable/src/afl-fuzz-bitmap.c#L481

but later on line https://github.com/AFLplusplus/AFLplusplus/blob/stable/src/afl-fuzz-bitmap.c#L491
it calls has_new_bits_unclassified.

post_exec should be called after computing the hash

**Comment by andreafioraldi** [comment]
mmm in the original afl and aflfast the hash is computed after classify_bits,https://github.com/mboehme/aflfast/blob/master/afl-fuzz.c#L2607

@vanhauser-thc do you remember why afl++ is computing the hash on the unclassified bitmap unlike afl?

**Comment by tokatoka** [comment]
https://github.com/AFLplusplus/AFLplusplus/commit/dacb2821b3a267b336b4465a24e1de22033ea279

This is the first commit that brought aflfast to afl++
here has_new_bit() is used

but later in 2020
it's changed to has_new_bits_unclassified
https://github.com/AFLplusplus/AFLplusplus/commit/d07b0169cb6137e4620db63b144e46e140b7437d

**Comment by tokatoka** [comment]
>  in general makes sense to hash the classified bitmap IMO

but unclassfied one should have richer info so isn't unclassified hash better?

**Comment by domenukk** [comment]
@andreafioraldi I feel like hitcounts is different to other observers, as in, it's not a raw observation but changes the observation state. A bit odd / doesn't perfectly fit.

File: libafl/src/schedulers/accounting.rs [hunk 3]
```diff
@@ -282,7 +277,7 @@ where
     }
 
     /// Creates a new [`CoverageAccountingScheduler`] that wraps a `base` [`Scheduler`]
-    /// and has a default probability to skip non-faved [`Testcase`]s of [`DEFAULT_SKIP_NON_FAVORED_PROB`].
+    /// and has a default probability to skip non-faved Testcases of [`DEFAULT_SKIP_NON_FAVORED_PROB`].
```
Comment (domenukk) [inline libafl/src/schedulers/accounting.rs:280]: Why remove the link?

**Review by domenukk — APPROVED** [review]

File: libafl/src/schedulers/accounting.rs (hunk 3 above)
Comment (tokatoka) [reply libafl/src/schedulers/accounting.rs:280]: `Testcase` is not used in Rust code (but only in this comment) so I deleted the `use` declaractions.

File: libafl/src/schedulers/mod.rs (hunk 2 above)
Comment (tokatoka) [reply libafl/src/schedulers/mod.rs:124]: schedulers change the data of testcases when on_add, on_next... happens. so I think it's better to put them together.

**Comment by vanhauser-thc** [comment]
honstely I do not remember about classified/unclassified map used for hashing.
I think classified makes more sense, but due a performance patch that brought enhanced classification in multiple steps that might be the reason that changed.

**Comment by andreafioraldi** [comment]
I think I will run a fuzzbench local run of afl++ with classified and unclassified soon

**Comment by tokatoka** [comment]
> I think I will run a fuzzbench local run of afl++ with classified and unclassified soon

please make sure to use PowerSchedule::FAST in fuzzbench 😜 
it's explore now

**Comment by vanhauser-thc** [comment]
> I think I will run a fuzzbench local run of afl++ with classified and unclassified soon

oh yes please. curious about the results.

File: libafl/src/schedulers/accounting.rs (hunk 3 above)
Comment (domenukk) [reply libafl/src/schedulers/accounting.rs:280]: Ok, fair

**Comment by domenukk** [comment]
Let's merge this?

**Comment by andreafioraldi** [comment]
> Let's merge this?

no, after the experiment

**Review by andreafioraldi — CHANGES_REQUESTED** [review]
on_exec/on_eval must be propagated to the inner scheduler in the coverage accounting scheduler too

**Comment by domenukk** [comment]
I don't care for this part of the PR, but I'd need scheduled_count and the parent id stuff

**Review by andreafioraldi — APPROVED** [review]

**Pull request merged**